### ISSUE TITLE
Make Ceylon scriptable

### ISSUE THREAD

**Issue by FroMage** [ceylon/ceylon-spec#200]
Needs some thinking

**Comment by gavinking** [comment]
Since @zoek1 is going to start work on this issue, I think it's time to finally make some decisions on this issue.

I propose that:

- a _script_ is a sort of lightweight module definition,
- scripts don't have any relationships to module repos, nor source folders, etc, they can sit wherever is convenient
- scripts _can_ be compiled (to a `.car` or `.js` in a module repo), but usually we just distribute them as plain `.ceylon` source files


The syntax of a script is:

1. a `module` statement,
2. optionally, one or more module `import` statements,
3. optionally, one or more package `import` statement,
4. a list of statements and declarations, interpreted as if they occur in the body of a toplevel function.

For example:

```ceylon
module script;
import ceylon.collection "1.1.1";
import ceylon.collection { ArrayList }

value list = ArrayList { "hello", "world" };
printAll(list);
```

An open question is whether we allow a script to be split across multiple `.ceylon` files, or if we force you to put everything in one source file.

**Comment by lucaswerkmeister** [comment]
Why is the `module` statement required?

**Comment by gavinking** [comment]
@lucaswerkmeister So that the parser can distinguish that this is a script.

**Comment by gavinking** [comment]
P.S. I'm OK with allowing just:

```ceylon
module;
print("hello world");
```

if you're happy for your script to execute in the default module.

**Comment by lucaswerkmeister** [comment]
So a module statement is
```antlr
ModuleStatement: Annotations "module" FullPackageName? ";"
```
and everything else is unchanged, except that module imports are now toplevel. (Fine by me, feels a bit more “script-y” than wrapping them in the module descriptor.)

:+1:

On semantics: I assume the typechecker checks the script just like a function body, in that “toplevel” declarations must work top-down and can’t be circular?

**Comment by gavinking** [comment]
> I assume the typechecker checks the script just like a function body, in that “toplevel” declarations must work top-down and can’t be circular?

Precisely. I briefly thought of making it work like the body of a class (initializer than declarations) but then I realized that there would be no point, since there's nothing to _call_ the declarations.

**Comment by quintesse** [comment]
But we'll need to "enforce" the uniqueness requirement of the module name somehow, right?
Especially when you allow them to be default modules as well.

So probably the real question is: where do we store the .cars for those compiled scripts?

Languages like python will create a `foo.pyc` compiled file beside the `foo.py` file, so we could create an `exampl.car` next to the `example.ceylon` script. Of course that would require you to always have write access to the folder where the source file is located.
(It does make "clean-up" easier, if you remove the source file you probably won't forget to delete the compiled version.)

We could of course just compile them to `./modules` as we do now but I have the feeling people don't like to see their (non-ceylon) folders littered with seemingly unrelated files/folders. A variant of this would be to compile scripts to `./.ceylon/modules` effectively hiding the output folder from sight. (Clean-up is more difficult though in these cases.)

We could also compile them to a fixed place like the cache (`~/.ceylon/cache`) but we don't allow default modules there (and how to distinguish between them?) so that's not really an option.

Another option would be a new folder, something like `~/.ceylon/compiled-scripts` and then match to the exact location of the script on disk somehow (eg. `~/.ceylon/compiled-scripts/home/user/ceylon/bin/example.ceylon/default.car`).

(The problem with those last two options is also the clean-up , they might fill up with crap pretty quickly if you use a lot of scripts just a couple of times. The folder itself is pretty easy to delete though.)

Personally I think the first option of generating the .car next to the script file is the most flexible, perhaps with a fallback to the `~/.ceylon/compiled-scripts` option when you don't have write access to the folder.

**Comment by quintesse** [comment]
Oh btw, as side-effect of the above, where compiled scripts are not really stored in the normal modules repositories, I think scripts should always be default modules: they can import but can never be imported themselves.

**Comment by quintesse** [comment]
Unless we also want the script source syntax available for normal Ceylon modules, as this really quick to write, single-file module. That would be great, but it would then have to follow all the normal Ceylon source rules: be in a `source` folder, be compiled with `ceylon compile` etc.
In the above messages I'm referring to Ceylon scripts as used in system scripting, being able to do:

```ceylon
#!/bin/ceylon run

module;
print("hello world"):
```

**Comment by gavinking** [comment]
> Oh btw, as side-effect of the above, where compiled scripts are not really stored in the normal modules repositories, I think scripts should always be default modules: they can import but can never be imported themselves.

I disagree, and I don't see how it follows.

**Comment by quintesse** [comment]
> I disagree, and I don't see how it follows.

Well if they're not part of the normal module repositories how will they be found?

(Unless they can only be imported by other scripts in the same folder.)

**Comment by PhiLhoSoft** [comment]
"_An open question is whether we allow a script to be split across multiple .ceylon files_"
I guess this will be requested sooner or later, to allow using script libraries and for readability / maintenance / reusability.
Even a small scripting language like [Lua](http://www.lua.org/manual/5.3/manual.html) allows this, via dofile() or the more formal require().

**Comment by quintesse** [comment]
> "An open question is whether we allow a script to be split across multiple .ceylon files"
> I guess this will be requested sooner or later

I'm pretty sure you're right people will request this, but I'm not sure *we* should go that way. I'd rather see the scripting files as the "glue" for proper modules. So they're really easy to use and fast to write needing no setup of folder structures or anything like that and maybe even perfect for embedding. But the moment you start re-using code from other scripts IMO you're far better of writing a proper module for that.

An the other hand I think I wouldn't mind too much if it was a C-like "#include" at the level of the parser or something like that. Something that would only be allowed in script files. (for ma affirming the whole idea that they're nameless and module-less)

**Comment by PhiLhoSoft** [comment]
That's the dofile() way, then... It does a bit more than #include, but that's the idea, without having to manage paths, versions, dependencies, etc.

And indeed, if the need for reusability arises, one can always use "real" Ceylon code for that.

**Comment by gavinking** [comment]
> Well if they're not part of the normal module repositories how will they be found?

Well the _compiled_ artifacts would go into repos. But the `script.ceylon` file would be specified on the command line.

**Comment by quintesse** [comment]
> Well the compiled artifacts would go into repos

But which repository? If a script file can be anywhere on your system? (Unless you need to specify it with an option, but that doesn't seem like a good default)
And what about having several script files all being nameless (going into the default module)? How do we distinguish them?

**Comment by jvasileff** [comment]
> Another option would be a new folder, something like ~/.ceylon/compiled-scripts and then match to the exact location of the script on disk somehow (eg. ~/.ceylon/compiled-scripts/home/user/ceylon/bin/example.ceylon/default.car).

That sounds compelling, but I can imagine a *ton* of ways it could go wrong; filesystem paths are far from straightforward and reliable.

So, what about using a shasum of the script file's contents, like `~/.ceylon/compiled-scripts/db32869dcc7cc6762af350447c0a19037792c5eb.car`? Perhaps a sanitized version of the script's filename could be included as a prefix to make the filenames somewhat human readable.

> Personally I think the first option of generating the .car next to the script file is the most flexible, perhaps with a fallback to the ~/.ceylon/compiled-scripts option when you don't have write access to the folder.

I agree. That seems like the safest and easiest to use approach, despite the littering of the filesystem. The `car` should be hidden and named with a pattern that is easy to add to an ignore file.

**Comment by quintesse** [comment]
> So, what about using a shasum of the script file's contents

That's certainly an option, but imagine the mess that will make *while developing*, you'd get a new version for each change you make.

> but I can imagine a ton of ways it could go wrong

What ways do you see this going wrong? One of the things I could think off is that you might hit filesystem limits (trying to create a deeply nested folder inside a deeply nested folder for example).
But maybe an option is to use not an shasum of the code but of the path?

**Comment by jvasileff** [comment]
> new version for each change you make

Good point.

> What ways do you see this going wrong?

Well, perhaps it's a solvable problem, but my inclination is always to avoid using user-input as filenames, if possible.

I'm thinking of length limits, as you mentioned, and also issues with canonicalization of unicode characters in filenames across filesystems of different types (see `rsync --iconv`), different filename rules for different filesystem types (extfs, hfs, ntfs, fat, etc.), translations performed by network filesystems (nfs, smb, etc), possible bugs determining path separators, symlink normalizations, unstable or reusable mount points, especially for removable media, and who knows what else.

Some of this may be FUD, but like I said, it's a mess I usually try to avoid.

> shasum of the code but of the path

That sounds interesting. For safety, I think it would make sense to validate the shasum of the code as well (somehow) when the `car` is not stored in the same directory. Timestamps may not be a sufficient guarantee that you are actually running the right binary.

Generally, I think it's wrong to trust a path as a universal identifier. Perhaps things like concat(filesystem-guid, inode) when supported work, but even that's defeated by cloning a disk with `dd`.

**Comment by quintesse** [comment]
> I'm thinking of ...

Ok, that's indeed all true, so indeed (as expected) trying to recreate the path would probably cause all kinds of headaches. So the shasum of the final URI or something like that would probably be enough. That you won't be able to easily match a script name to a file in the cache I don't really care that much about. If really needed we could probably make it an option of the script command to tell you where the compiled version is located.

>  I think it would make sense to validate the shasum of the code as well

It's certainly a possibility, but I'm not sure it's worth it, somebody would really have to jump through some hoops to get the wrong code to execute it seems to me.

**Comment by HenningB** [comment]
I'd prefer not to base the shasum on the path/URI/URL. One pattern I use whilst developing/hacking is to copy a file or directory (as backup), experiment with the original, throw it away, and then renaming the backup to the original name. Needing to remember to clean the script-cache would lead to confusion.

**Comment by jvasileff** [comment]
Could use the sha of the file contents for the car file name/PK, and the path as part of a cache eviction strategy.



On Apr 28, 2015, at 2:01 PM, Henning Burdack <notifications@github.com<mailto:notifications@github.com>> wrote:


I'd prefer not to base the shasum on the path/URI/URL. One pattern I use whilst developing/hacking is to copy a file or directory (as backup), experiment with the original, throw it away, and then renaming the backup to the original name. Needing to remember to clean the script-cache would lead to confusion.

-

**Comment by quintesse** [comment]
> and the path as part of a cache eviction strategy.

How that would work? For me the opposite would seem to be the more logical way: to find it you'd use the hashed URI/Path and then to know if you'd need to update it you could look at the hashed contents.

@HenningB if we don't go for option 1 (local next to source) + option 5 (shasummed paths) as a fallback, but solely for option 5 then that might be a problem yes.

**Comment by jvasileff** [comment]
I guess you could use `$ContentsHash-$PathHash.car` which would allow either one-to-many or one-to-one binary-to-sourceFile, depending on which way you want to go. (In the one-to-many scenario, *executing* wouldn't necessarily require a match on the `$PathHash` part).

**Comment by lucono** [comment]
> That sounds interesting. For safety, I think it would make sense to validate the shasum of the code as well (somehow) when the car is not stored in the same directory. Timestamps may not be a sufficient guarantee that you are actually running the right binary.

I agree with this opinion, but even also for when the file is stored in the same directory, as a way of verifying that it has not been updated since it was last compiled.

> > Could use the sha of the file contents for the car file name/PK, and the path as part of a cache eviction strategy.
> 
> How that would work? For me the opposite would seem to be the more logical way: to find it you'd use the hashed URI/Path and then to know if you'd need to update it you could look at the hashed contents.

This makes sense to me.

Even if storing the compiled .car locally next to the script file, I think it would make sense to compare a fresh hash of the script contents to the hash of the original script from which the existing .car was compiled, to know if re-compilation is required prior to execution.

For scripts stored in the scripts cache location because the user does not have write access to the local directory of the script, some form of the hash of the URI/path of the script file could be used as already proposed, to locate the right .car in the cache, after which the hash of the script contents would still need to be compared to determine if re-compilation is required prior to execution.

Could be something like:

``` sh
~/.ceylon/compiled-scripts/db32869dcc7cc6762af350447c0a19037792c5eb.car
~/.ceylon/compiled-scripts/db32869dcc7cc6762af350447c0a19037792c5eb.md5
```

Unless it's possible to also bundle the hash of the original source script into the .car file, making the separate .md5 file unnecessary.
